___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___

# Exploratory Data Analysis Project (Part - 1) - Session - 3

<blockquote style="background: #f9f9f9;padding: 0.5em 10px;" class="">
<img src="https://docs.google.com/uc?id=1FOcuE9AkAqc0JU6dUly-cuoUg3bOp92H" class="img-fluid" width="200" height="85"> <h1>EDA for Car Price Prediction Model</h1>
</blockquote>

In [1]:
import pandas as pd
import numpy as np
import warnings 

pd.set_option('display.max_columns', 500)
pd.set_option('display.max_colwidth', 500)
pd.set_option('display.max_info_columns', 500)
pd.set_option('display.max_info_rows', 2000)
pd.set_option('display.expand_frame_repr', True)
pd.set_option('display.width', 2000)

warnings.filterwarnings("ignore")
warnings.warn("this will not show")

In [2]:
oto = pd.read_json("scout_car.json", orient = "records", lines = True)

In [3]:
oto.head()

url make_model                                   short_description body_type  price               vat         km registration         prev_owner  kW      hp                                     Type Previous Owners                        Next Inspection                      Inspection new                          Warranty                      Full Service   Non-smoking Vehicle null      Make         Model     Offer Number First Registration       Body Color      Paint Type           Body Color Original          Upholstery              Body Nr. of Doors Nr. of Seats      Model Code         Gearing Type    Displacement Cylinders          Weight  Drive chain                                   Fuel                                                               Consumption               CO2 Emission Emission Class                                                                                                                                                                                                                                                                                                     \nComfort & Convenience\n                                                                       \nEntertainment & Media\n                                                      \nExtras\n                                                                                                                                                                                                                                                                                 \nSafety & Security\n  \
0         https://www.autoscout24.com//offers/audi-a1-sportback-1-4-tdi-s-tronic-xenon-navi-klima-diesel-black-bdab349a-caa5-41b0-98eb-c1345b84445e    Audi A1         Sportback 1.4 TDI S-tronic Xenon Navi Klima    Sedans  15770    VAT deductible  56,013 km      01/2016  2 previous owners NaN   66 kW  [, Used, , Diesel (Particulate Filter)]           \n2\n  [\n06/2021\n, \n99 g CO2/km (comb)\n]               [\nYes\n, \nEuro 6\n]           [\n, \n, \n4 (Green)\n]                          [\n, \n]              [\n, \n]   []  \nAudi\n  [\n, A1, \n]  [\nLR-062483\n]     [\n, 2016, \n]  [\n, Black, \n]  [\nMetallic\n]           [\nMythosschwarz\n]  [\nCloth, Black\n]  [\n, Sedans, \n]      [\n5\n]      [\n5\n]  [\n0588/BDF\n]  [\n, Automatic, \n]  [\n1,422 cc\n]   [\n3\n]  [\n1,220 kg\n]  [\nfront\n]  [\n, Diesel (Particulate Filter), \n]  [[3.8 l/100 km (comb)], [4.3 l/100 km (city)], [3.5 l/100 km (country)]]   [\n99 g CO2/km (comb)\n]   [\nEuro 6\n]  [Air conditioning, Armrest, Automatic climate control, Cruise control, Electrical side mirrors, Hill Holder, Leather steering wheel, Light sensor, Multi-function steering wheel, Navigation system, Park Distance Control, Parking assist system sensors rear, Power windows, Rain sensor, Seat heating, Start-stop system]                                     [Bluetooth, Hands-free equipment, On-board computer, Radio]              [Alloy wheels, Catalytic Converter, Voice Control]                                          [ABS, Central door lock, Daytime running lights, Driver-side airbag, Electronic stability control, Fog lights, Immobilizer, Isofix, Passenger-side airbag, Power steering, Side airbag, Tire pressure monitoring system, Traction control, Xenon headlights]   
1                                      https://www.autoscout24.com//offers/audi-a1-1-8-tfsi-sport-gasoline-red-b2547f8a-e83f-6237-e053-e250040a56df    Audi A1                                      1.8 TFSI sport    Sedans  14500  Price negotiable  80,000 km      03/2017               None NaN  141 kW                     [, Used, , Gasoline]             NaN                                    NaN                                 NaN                               NaN                               NaN                   NaN   []  \nAudi\n  [\n, A1, \n]              NaN     [\n, 2017, \n]    [\n, Red, \n]             NaN                           NaN   [\nCloth, Grey\n]

##  Now let's try to examine our dataframe (oto)

In [4]:
oto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15919 entries, 0 to 15918
Data columns (total 54 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   url                            object 
 1   make_model                     object 
 2   short_description              object 
 3   body_type                      object 
 4   price                          int64  
 5   vat                            object 
 6   km                             object 
 7   registration                   object 
 8   prev_owner                     object 
 9   kW                             float64
 10  hp                             object 
 11  Type                           object 
 12  Previous Owners                object 
 13  Next Inspection                object 
 14  Inspection new                 object 
 15  Warranty                       object 
 16  Full Service                   object 
 17  Non-smoking Vehicle            object 
 18  null  

In [33]:
oto.describe()

price   kW
count  15919.000000  0.0
mean   18019.896727  NaN
std     7386.169409  NaN
min       13.000000  NaN
25%    12850.000000  NaN
50%    16900.000000  NaN
75%    21900.000000  NaN
max    74600.000000  NaN

## Handling with Missing Values.

- **Identifying the amount of the Missing Values**

In [5]:
oto.shape[0]

15919

In [6]:
oto.isnull().sum()*100/oto.shape[0]

url                                0.000000
make_model                         0.000000
short_description                  0.288963
body_type                          0.376908
price                              0.000000
vat                               28.349771
km                                 0.000000
registration                       0.000000
prev_owner                        42.892141
kW                               100.000000
hp                                 0.000000
Type                               0.012564
Previous Owners                   41.711163
Next Inspection                   77.793831
Inspection new                    75.299956
Warranty                          34.047365
Full Service                      48.395000
Non-smoking Vehicle               54.915510
null                               0.000000
Make                               0.000000
Model                              0.000000
Offer Number                      19.944720
First Registration              

In [36]:
col_list = list(oto.columns)

In [37]:
col_list

['url',
 'make_model',
 'short_description',
 'body_type',
 'price',
 'vat',
 'km',
 'registration',
 'prev_owner',
 'kW',
 'hp',
 'Type',
 'Previous Owners',
 'Next Inspection',
 'Inspection new',
 'Warranty',
 'Full Service',
 'Non-smoking Vehicle',
 'null',
 'Make',
 'Model',
 'Offer Number',
 'First Registration',
 'Body Color',
 'Paint Type',
 'Body Color Original',
 'Upholstery',
 'Body',
 'Nr. of Doors',
 'Nr. of Seats',
 'Model Code',
 'Gearing Type',
 'Displacement',
 'Cylinders',
 'Weight',
 'Drive chain',
 'Fuel',
 'Consumption',
 'CO2 Emission',
 'Emission Class',
 '\nComfort & Convenience\n',
 '\nEntertainment & Media\n',
 '\nExtras\n',
 '\nSafety & Security\n',
 'description',
 'Emission Label',
 'Gears',
 'Country version',
 'Electricity consumption',
 'Last Service Date',
 'Other Fuel Types',
 'Availability',
 'Last Timing Belt Service Date',
 'Available from']

## Let's begin to implement EDA Process to all *columns*

> ### We'll begin with the variables which has high percentage of missing values(NaNs). 
> ### We can remove some of them. At the begining we can focus on the columns that have %35 precentage and higher of missing values.

In [38]:
def show_nans(oto, limit):
    missing = oto.isnull().sum()*100/oto.shape[0]
    return missing.loc[lambda x : x > limit]
show_nans(oto, 35)

prev_owner                        42.892141
kW                               100.000000
Previous Owners                   41.711163
Next Inspection                   77.793831
Inspection new                    75.299956
Full Service                      48.395000
Non-smoking Vehicle               54.915510
Paint Type                        36.258559
Model Code                        68.729192
Cylinders                         35.680633
Weight                            43.809285
Drive chain                       43.080596
Emission Label                    74.967021
Country version                   52.346253
Electricity consumption           99.139393
Last Service Date                 96.444500
Other Fuel Types                  94.472015
Availability                      96.011056
Last Timing Belt Service Date     99.899491
Available from                    98.291350
dtype: float64

- We can drop the columns which have 90 percentage or high of missing values.

- ``'Availability'``, ``'Available from'``, ``'Electricity consumption'``, ``'Last Service Date'``, ``'Last Timing Belt Service Date'``, ``'Other Fuel Types'``, ``'kW'``

In [7]:
drop_list1 = ['Availability','Available from','Electricity consumption','Last Service Date','Last Timing Belt Service Date',
         'Other Fuel Types','kW']

In [8]:
oto.drop(drop_list1, axis=1, inplace=True)

``'prev_owner'`` is very similar to ``'Previous Owners'``.  ``'Weight'`` and ``'Non-smoking Vehicle'`` are assumed that they don't have meaningful  effect on the result. So we can drop them all.

### ``"prev_owner"`` and ``"Previous Owners"``

In [9]:
oto[["prev_owner", "Previous Owners"]]

prev_owner                                                                            Previous Owners
0      2 previous owners                                                                                      \n2\n
1                   None                                                                                        NaN
2       1 previous owner                                                                                      \n1\n
3       1 previous owner                                                                                      \n1\n
4       1 previous owner                                                                                      \n1\n
...                  ...                                                                                        ...
15914               None                                                                                        NaN
15915   1 previous owner  [\n1\n, \n, 7.4 l/100 km (comb), \n, 9.2 l/100 km (city), \n, 6.3 l/100 km (country), \n]
15916   1 previous owner                                                           [\n1\n, \n139 g CO2/km (comb)\n]
15917               None                                                                                        NaN
15918   1 previous owner                                                                                      \n1\n

[15919 rows x 2 columns]

#### These two columns are very similar to each other. Let's choose better of them.

In [10]:
oto["Previous Owners"]

0                                                                                            \n2\n
1                                                                                              NaN
2                                                                                            \n1\n
3                                                                                            \n1\n
4                                                                                            \n1\n
                                                   ...                                            
15914                                                                                          NaN
15915    [\n1\n, \n, 7.4 l/100 km (comb), \n, 9.2 l/100 km (city), \n, 6.3 l/100 km (country), \n]
15916                                                             [\n1\n, \n139 g CO2/km (comb)\n]
15917                                                                                          NaN
15918     

In [11]:
oto["prev_owner"]

0        2 previous owners
1                     None
2         1 previous owner
3         1 previous owner
4         1 previous owner
               ...        
15914                 None
15915     1 previous owner
15916     1 previous owner
15917                 None
15918     1 previous owner
Name: prev_owner, Length: 15919, dtype: object

In [12]:
oto["Previous Owners"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


\n1\n                                                                                        8101
NaN                                                                                          6640
\n2\n                                                                                         766
\n0\n                                                                                         163
\n3\n                                                                                          17
                                                                                             ... 
[\n1\n, \n96 g CO2/km (comb)\n]                                                                 1
[\n1\n, \n181 g CO2/km (comb)\n]                                                                1
[\n1\n, \n, 6.6 l/100 km (comb), \n, 8.2 l/100 km (city), \n, 5.6 l/100 km (country), \n]       1
[\n1\n, \n, 6.7 l/100 km (comb), \n, 8.6 l/100 km (city), \n, 5.6 l/100 km (country), \n]       1
[\n1\n, \n102 g CO2/

In [13]:
oto["prev_owner"].value_counts(dropna = False)

1 previous owner     8294
NaN                  6828
2 previous owners     778
3 previous owners      17
4 previous owners       2
Name: prev_owner, dtype: int64

### ``'Previous Owners'``

In [14]:
oto["Previous Owners"].str.strip("\n")

0          2
1        NaN
2          1
3          1
4          1
        ... 
15914    NaN
15915    NaN
15916    NaN
15917    NaN
15918      1
Name: Previous Owners, Length: 15919, dtype: object

In [15]:
oto["Previous Owners"].str.strip("\n").value_counts(dropna = False)

1      8101
NaN    6870
2       766
0       163
3        17
4         2
Name: Previous Owners, dtype: int64

In [16]:
oto["Previous Owners"].str[0]

0           \n
1          NaN
2           \n
3           \n
4           \n
         ...  
15914      NaN
15915    \n1\n
15916    \n1\n
15917      NaN
15918       \n
Name: Previous Owners, Length: 15919, dtype: object

In [17]:
oto['Previous Owners'].str[0].str[1]

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
15914    NaN
15915      1
15916      1
15917    NaN
15918    NaN
Name: Previous Owners, Length: 15919, dtype: object

In [18]:
oto['Previous Owners'].str[0].str[1].value_counts(dropna = False)

NaN    15689
1        193
0         25
2         12
Name: Previous Owners, dtype: int64

> ### We catch the values inside the lists.

> ### There are 230 new values which are not visible in previous format.

1   →     193

0   →     25

2   →     12

In [19]:
oto["Previous_Owner1"] = oto['Previous Owners'].str[0].str[1]
oto["Previous_Owner1"]

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
        ... 
15914    NaN
15915      1
15916      1
15917    NaN
15918    NaN
Name: Previous_Owner1, Length: 15919, dtype: object

In [20]:
oto['Previous_Owner'] = oto['Previous Owners'].str.strip('\n')
oto['Previous_Owner']

0          2
1        NaN
2          1
3          1
4          1
        ... 
15914    NaN
15915    NaN
15916    NaN
15917    NaN
15918      1
Name: Previous_Owner, Length: 15919, dtype: object

In [21]:
oto.Previous_Owner.fillna(oto.Previous_Owner1, inplace=True)
oto.Previous_Owner.value_counts(dropna=False) 

1      8294
NaN    6640
2       778
0       188
3        17
4         2
Name: Previous_Owner, dtype: int64

### ``"Previous_Owner"`` our new column. Now, we have to fillna this column in the most convenient way

### We can fill the missing values with proportion of other existing values using `bfill` method.

In [22]:
oto["Previous_Owner"].fillna(method='bfill', inplace=True) 

In [23]:
oto["Previous_Owner"].value_counts(dropna = False)

1    14169
2     1179
0      531
3       38
4        2
Name: Previous_Owner, dtype: int64

### ``"prev_owner"``

In [24]:
oto.prev_owner.value_counts(dropna = False)

1 previous owner     8294
NaN                  6828
2 previous owners     778
3 previous owners      17
4 previous owners       2
Name: prev_owner, dtype: int64

In [25]:
oto["prev_owner"] = oto["prev_owner"].str.replace(" previous owner", "").str.replace("s", "")

In [26]:
oto.prev_owner.value_counts(dropna = False)

1      8294
NaN    6828
2       778
3        17
4         2
Name: prev_owner, dtype: int64

In [27]:
oto["prev_owner"].fillna(method='bfill', inplace=True)

In [28]:
oto.prev_owner.value_counts(dropna = False)

1    14692
2     1187
3       38
4        2
Name: prev_owner, dtype: int64

In [29]:
oto["Previous_Owner"].value_counts()

1    14169
2     1179
0      531
3       38
4        2
Name: Previous_Owner, dtype: int64

**It's better to drop ``"prev_owner"`` column in order to benefit more from ``"Previous_Owner"``. By the way, It is time to drop the other useless columns : `"Weight"`, `'Non-smoking Vehicle'`.**

In [30]:
drop_list2 = ["prev_owner", 'Weight','Non-smoking Vehicle']

In [31]:
oto.drop(drop_list2, axis = 1, inplace = True)

### ``'body_type'`` is also very similar to ``'Body'``.

- **Defining a Function**

In [35]:
def fill_most(oto, group_col, col_name):
    for group in list(oto[group_col].unique()) :
        grp_inx = list(oto[oto[group_col] == group][col_name].index)
        oto[col_name].iloc[grp_inx] = oto[col_name].iloc[grp_inx].fillna(oto[oto[group_col] == group][col_name].mode()[0])

In [ ]:
#fill_most(oto, "make_model", "body_type")

In [ ]:
#fill_most(oto, "make_model", "Extras")

In [ ]:
#fill_most(oto, "make_model", "Comfort&Convenience")

In [19]:
#oto["make_model"].unique()

array(['Audi A1', 'Audi A2', 'Audi A3', 'Opel Astra', 'Opel Corsa',
       'Opel Insignia', 'Renault Clio', 'Renault Duster',
       'Renault Espace'], dtype=object)

In [51]:
oto["body_type"].iloc[grp_inx]

2615      Convertible
2616           Sedans
2617           Sedans
2618           Sedans
2619           Sedans
            ...      
5707           Sedans
5708           Sedans
5709    Station wagon
5710           Sedans
5711           Sedans
Name: body_type, Length: 3097, dtype: object

In [53]:
oto[oto["make_model"] == "Audi A3"]["body_type"]

2615      Convertible
2616           Sedans
2617           Sedans
2618           Sedans
2619           Sedans
            ...      
5707           Sedans
5708           Sedans
5709    Station wagon
5710           Sedans
5711           Sedans
Name: body_type, Length: 3097, dtype: object

In [39]:
grp_inx = list(oto[oto["make_model"] == "Audi A3"]["body_type"].index)

In [45]:
grp_inx

[2615,
 2616,
 2617,
 2618,
 2619,
 2620,
 2621,
 2622,
 2623,
 2624,
 2625,
 2626,
 2627,
 2628,
 2629,
 2630,
 2631,
 2632,
 2633,
 2634,
 2635,
 2636,
 2637,
 2638,
 2639,
 2640,
 2641,
 2642,
 2643,
 2644,
 2645,
 2646,
 2647,
 2648,
 2649,
 2650,
 2651,
 2652,
 2653,
 2654,
 2655,
 2656,
 2657,
 2658,
 2659,
 2660,
 2661,
 2662,
 2663,
 2664,
 2665,
 2666,
 2667,
 2668,
 2669,
 2670,
 2671,
 2672,
 2673,
 2674,
 2675,
 2676,
 2677,
 2678,
 2679,
 2680,
 2681,
 2682,
 2683,
 2684,
 2685,
 2686,
 2687,
 2688,
 2689,
 2690,
 2691,
 2692,
 2693,
 2694,
 2695,
 2696,
 2697,
 2698,
 2699,
 2700,
 2701,
 2702,
 2703,
 2704,
 2705,
 2706,
 2707,
 2708,
 2709,
 2710,
 2711,
 2712,
 2713,
 2714,
 2715,
 2716,
 2717,
 2718,
 2719,
 2720,
 2721,
 2722,
 2723,
 2724,
 2725,
 2726,
 2727,
 2728,
 2729,
 2730,
 2731,
 2732,
 2733,
 2734,
 2735,
 2736,
 2737,
 2738,
 2739,
 2740,
 2741,
 2742,
 2743,
 2744,
 2745,
 2746,
 2747,
 2748,
 2749,
 2750,
 2751,
 2752,
 2753,
 2754,
 2755,
 2756,
 2757,

In [50]:
oto["body_type"].iloc[2680]

'Sedans'

In [ ]:
# oto[col_name].iloc[grp_inx] = oto[col_name].iloc[grp_inx].fillna()

In [44]:
# oto[oto["make_model"] == "Audi A3"]["body_type"].fillna("Sedans")

2615      Convertible
2616           Sedans
2617           Sedans
2618           Sedans
2619           Sedans
            ...      
5707           Sedans
5708           Sedans
5709    Station wagon
5710           Sedans
5711           Sedans
Name: body_type, Length: 3097, dtype: object

In [26]:
oto[oto["make_model"] == "Audi A3"]["body_type"].mode()[0]

'Sedans'

In [27]:
oto[oto["make_model"] == "Audi A1"]["body_type"].mode()[0]

'Sedans'

In [28]:
oto[oto["make_model"] == "Renault Espace"]["body_type"].mode()[0]

'Van'

***
***

### ``'body_type'``

In [32]:
oto["body_type"].value_counts(dropna = False)

Sedans           7903
Station wagon    3553
Compact          3153
Van               783
Other             290
Transporter        88
NaN                60
Off-Road           56
Coupe              25
Convertible         8
Name: body_type, dtype: int64

In [34]:
oto.groupby("make_model")["body_type"].value_counts(dropna = False)

make_model      body_type    
Audi A1         Sedans           1538
                Compact          1039
                Station wagon      21
                Other              13
                Coupe               2
                Van                 1
Audi A2         Off-Road            1
Audi A3         Sedans           2598
                Station wagon     282
                Compact           182
                Other              16
                Convertible         8
                NaN                 7
                Coupe               4
Opel Astra      Station wagon    1211
                Sedans           1053
                Compact           185
                Other              67
                NaN                 7
                Coupe               2
                Off-Road            1
Opel Corsa      Compact          1230
                Sedans            875
                Other              87
                Coupe              13
                Tran

In [36]:
fill_most(oto, "make_model", "body_type")

In [37]:
oto.body_type.value_counts(dropna = False)

Sedans           7925
Station wagon    3563
Compact          3155
Van               809
Other             290
Transporter        88
Off-Road           56
Coupe              25
Convertible         8
Name: body_type, dtype: int64

### ``"Body"``

In [33]:
oto["Body"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n, Sedans, \n]           7903
[\n, Station wagon, \n]    3553
[\n, Compact, \n]          3153
[\n, Van, \n]               783
[\n, Other, \n]             290
[\n, Transporter, \n]        88
NaN                          60
[\n, Off-Road, \n]           56
[\n, Coupe, \n]              25
[\n, Convertible, \n]         8
Name: Body, dtype: int64

### It seems that ``"Body"`` deserves to be dropped

In [38]:
oto.drop("Body", axis = 1, inplace = True)

***

### ``'Inspection new'``

In [39]:
oto["Inspection new"].isnull().sum()/len(oto["Inspection new"])

0.7529995602738866

#### It has high percentage of missing values, but it is important feature for the result (predicting price). We should keep it for now and we can assume missing values as 'no',  so that we can fillna missing values with string ``'no'``.

In [40]:
oto["Inspection new"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                            11987
[\nYes\n, \nEuro 6\n]                                                                            523
\nYes\n                                                                                          362
[\nYes\n, \n102 g CO2/km (comb)\n]                                                               174
[\nYes\n, \n4 (Green)\n]                                                                         166
                                                                                               ...  
[\nYes\n, \n1 (No sticker)\n]                                                                      1
[\nYes\n, \n, 5.2 l/100 km (comb), \n, 6.6 l/100 km (city), \n, 4.3 l/100 km (country), \n]        1
[\nYes\n, \n, 5.7 l/100 km (comb), \n, 7.3 l/100 km (city), \n, 4.9 l/100 km (country), \n]        1
[\nYes\n, \n, 5.2 l/100 km (comb), \n, 6.8 l/100 km (city), \n, 4.3 l/100 km (country), \n]

In [41]:
oto['Inspection new'].str[0].value_counts(dropna = False)

NaN        11987
\nYes\n     3570
\n           362
Name: Inspection new, dtype: int64

#### We chose the ``'\n'`` string in ``'.contains()'``. So we cover both variables in the lists and out of the lists
#### ``'*'`` 1 → gives us 1 or 0 according to True/False

In [42]:
oto['Inspection new'].str[0].str.contains("\n", na = False)

0         True
1        False
2        False
3        False
4         True
         ...  
15914    False
15915    False
15916     True
15917    False
15918     True
Name: Inspection new, Length: 15919, dtype: bool

In [43]:
True * 1

1

In [44]:
False * 1

0

In [45]:
oto["Inspection_New_Bool"] = oto['Inspection new'].str[0].str.contains("\n", na = False)*1

In [46]:
oto["Inspection_New_Bool"].value_counts(dropna = False)

0    11987
1     3932
Name: Inspection_New_Bool, dtype: int64

***

### ``'Next Inspection'``

In [47]:
oto['Next Inspection'].isnull().sum()/len(oto['Next Inspection'])

0.7779383127080847

In [54]:
oto["Next Inspection"]

0         [\n06/2021\n, \n99 g CO2/km (comb)\n]
1                                           NaN
2                                           NaN
3                                           NaN
4                                           NaN
                          ...                  
15914                                       NaN
15915    [\n01/2022\n, \n168 g CO2/km (comb)\n]
15916                                       NaN
15917                                       NaN
15918                                       NaN
Name: Next Inspection, Length: 15919, dtype: object

In [48]:
oto["Next Inspection"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                       12384
\n04/2022\n                                  62
\n03/2021\n                                  38
\n03/2022\n                                  36
\n06/2021\n                                  34
                                          ...  
[\n07/2021\n, \n116 g CO2/km (comb)\n]        1
[\n08/2021\n, \n110 g CO2/km (comb)\n]        1
[\n02/2021\n, \n149 g CO2/km (comb)\n]        1
[\n11/2021\n, \n161 g CO2/km (comb)\n]        1
[\n02/2022\n, \n114 g CO2/km (comb)\n]        1
Name: Next Inspection, Length: 1400, dtype: int64

In [49]:
a = ["I am a list"]

In [50]:
type(a) == list

True

In [51]:
"".join(a).strip()

'I am a list'

In [52]:
nxtins_list = ["".join(item).strip() if type(item) == list else item for item in oto['Next Inspection']]
oto['next_inspection'] = pd.DataFrame(nxtins_list)

In [53]:
oto.next_inspection.value_counts(dropna = False)

NaN                                                                            12384
\n04/2022\n                                                                       62
\n03/2021\n                                                                       38
\n03/2022\n                                                                       36
\n06/2021\n                                                                       34
                                                                               ...  
06/2021\n\n6.1 l/100 km (comb)\n7.7 l/100 km (city)\n5.2 l/100 km (country)        1
12/2019\n\n139 g CO2/km (comb)                                                     1
11/2019\n\n149 g CO2/km (comb)                                                     1
04/2021\n\n101 g CO2/km (comb)                                                     1
05/2019\n\n4.9 l/100 km (comb)\n6.2 l/100 km (city)\n4.2 l/100 km (country)        1
Name: next_inspection, Length: 1400, dtype: int64

In [54]:
oto["next_inspection"] = oto.next_inspection.str.extract("(\d\d/\d\d\d\d)")

In [55]:
oto.next_inspection.value_counts(dropna = False)

NaN        12384
06/2021      471
03/2021      210
05/2021      180
04/2021      171
           ...  
11/2022        1
05/2014        1
04/2016        1
01/1955        1
03/2024        1
Name: next_inspection, Length: 78, dtype: int64

In [56]:
oto.next_inspection = pd.to_datetime(oto.next_inspection, format = "%m/%Y")

In [57]:
oto.next_inspection.value_counts(dropna=False)

NaT           12384
2021-06-01      471
2021-03-01      210
2021-05-01      180
2021-04-01      171
              ...  
2016-04-01        1
1955-01-01        1
1999-01-01        1
2017-05-01        1
2016-05-01        1
Name: next_inspection, Length: 78, dtype: int64

### `"next_inspection"` has %77 percent of missing value. For now, let's leave it as it is. We can take a look at it again later.

### ``'Paint Type'``

In [58]:
oto["Paint Type"].isnull().sum()/len(oto["Paint Type"])*100

36.25855895470821

In [59]:
oto["Paint Type"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\nMetallic\n]       9794
NaN                  5772
[\nUni/basic\n]       347
[\nPerl effect\n]       6
Name: Paint Type, dtype: int64

In [60]:
oto["Paint Type"].str[0]

0        \nMetallic\n
1                 NaN
2        \nMetallic\n
3        \nMetallic\n
4        \nMetallic\n
             ...     
15914    \nMetallic\n
15915    \nMetallic\n
15916             NaN
15917             NaN
15918    \nMetallic\n
Name: Paint Type, Length: 15919, dtype: object

In [62]:
oto["Paint Type"].str[0].str.strip("\n").value_counts(dropna = False)

Metallic       9794
NaN            5772
Uni/basic       347
Perl effect       6
Name: Paint Type, dtype: int64

### We should decide or specify its importance or value level for price prediction.

In [63]:
oto['paint_type'] = oto['Paint Type'].str[0].str.strip('\n')

In [64]:
oto['paint_type'].value_counts(dropna = False)

Metallic       9794
NaN            5772
Uni/basic       347
Perl effect       6
Name: paint_type, dtype: int64

***

### ``'Country version'``

In [65]:
oto["Country version"].isnull().sum()/len(oto["Country version"])

0.5234625290533325

### It seems more than half of the data are missing.

In [66]:
oto["Country version"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                     8333
[\nGermany\n]           4502
[\nItaly\n]             1038
[\nEuropean Union\n]     507
[\nNetherlands\n]        464
[\nSpain\n]              325
[\nBelgium\n]            314
[\nAustria\n]            208
[\nCzech Republic\n]      52
[\nPoland\n]              49
[\nFrance\n]              38
[\nDenmark\n]             33
[\nHungary\n]             28
[\nJapan\n]                8
[\nSlovakia\n]             4
[\nCroatia\n]              4
[\nSweden\n]               3
[\nRomania\n]              2
[\nBulgaria\n]             2
[\nSerbia\n]               1
[\nLuxembourg\n]           1
[\nSwitzerland\n]          1
[\nSlovenia\n]             1
[\nEgypt\n]                1
Name: Country version, dtype: int64

In [67]:
oto['Country version'].str[0].str.strip('\n').value_counts(dropna=False) 

NaN               8333
Germany           4502
Italy             1038
European Union     507
Netherlands        464
Spain              325
Belgium            314
Austria            208
Czech Republic      52
Poland              49
France              38
Denmark             33
Hungary             28
Japan                8
Slovakia             4
Croatia              4
Sweden               3
Bulgaria             2
Romania              2
Egypt                1
Slovenia             1
Luxembourg           1
Serbia               1
Switzerland          1
Name: Country version, dtype: int64

### It's not essential for the result (price) but we can keep it.
### May be we can fillna using weight-mean in terms of values respectively

In [68]:
oto['country'] = oto['Country version'].str[0].str.strip('\n')

***

### ``'Cylinders'``

In [69]:
oto.Cylinders.isnull().sum()/len(oto.Cylinders)

0.3568063320560337

In [70]:
oto.Cylinders.value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[\n4\n]    8105
NaN        5680
[\n3\n]    2104
[\n5\n]      22
[\n6\n]       3
[\n2\n]       2
[\n8\n]       2
[\n1\n]       1
Name: Cylinders, dtype: int64

In [71]:
oto['Cylinders'].str[0].str.strip('\n').value_counts(dropna=False) 

4      8105
NaN    5680
3      2104
5        22
6         3
8         2
2         2
1         1
Name: Cylinders, dtype: int64

### We can handle with the missing values later. Let's keep it for now.

In [72]:
oto['Cylinders'] = oto['Cylinders'].str[0].str.strip('\n')

***

### ``'Emission Label'``

In [73]:
oto['Emission Label'].isnull().sum()/len(oto['Emission Label'])

0.749670205414913

### It has very high percentage of missing values. The percent of the missing values is high enough to drop but let's keep it for now.

In [74]:
oto["Emission Label"].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                     11934
[\n4 (Green)\n]          3553
[\n1 (No sticker)\n]      381
[[], [], []]               40
[\n5 (Blue)\n]              8
[\n3 (Yellow)\n]            2
[\n2 (Red)\n]               1
Name: Emission Label, dtype: int64

In [75]:
oto['Emission Label'].str[0].str.strip('\n').value_counts(dropna=False) 

NaN               11974
4 (Green)          3553
1 (No sticker)      381
5 (Blue)              8
3 (Yellow)            2
2 (Red)               1
Name: Emission Label, dtype: int64

In [76]:
oto['emission_label'] = oto['Emission Label'].str[0].str.strip('\n')

In [77]:
oto.emission_label.value_counts(dropna = False)

NaN               11974
4 (Green)          3553
1 (No sticker)      381
5 (Blue)              8
3 (Yellow)            2
2 (Red)               1
Name: emission_label, dtype: int64

***

### ``'Full Service'``

In [80]:
oto['Full Service'].isnull().sum()/len(oto['Full Service'])*100

48.394999685909916

In [78]:
oto['Full Service'].value_counts(dropna=False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                           7704
[\n, \n, \n4 (Green)\n]                                                                       2235
[\n, \n, \nEuro 6\n]                                                                          2097
[\n, \n]                                                                                      1702
[\n, \n, \nEuro 6d-TEMP\n]                                                                     399
                                                                                              ... 
[\n, \n, \n, 6 l/100 km (comb), \n, 7.5 l/100 km (city), \n, 5.2 l/100 km (country), \n]         1
[\n, \n, \n, 5.5 l/100 km (comb), \n, 7.2 l/100 km (city), \n, 4.5 l/100 km (country), \n]       1
[\n, \n, \n, 5.8 l/100 km (comb), \n, 7.4 l/100 km (city), \n, 4.9 l/100 km (country), \n]       1
[\n, \n, \n, 5.4 l/100 km (comb), \n, 7 l/100 km (city), \n, 4.5 l/100 km (country), \n]         1
[\n, \n, \

In [81]:
list(oto['Full Service'].value_counts(dropna=False).index)

[nan,
 ['\n', '\n', '\n4 (Green)\n'],
 ['\n', '\n', '\nEuro 6\n'],
 ['\n', '\n'],
 ['\n', '\n', '\nEuro 6d-TEMP\n'],
 '\n',
 ['\n', '\n', '\n120 g CO2/km (comb)\n'],
 ['\n', '\n', '\n126 g CO2/km (comb)\n'],
 ['\n', '\n', '\n104 g CO2/km (comb)\n'],
 ['\n', '\n', '\n97 g CO2/km (comb)\n'],
 ['\n', '\n', '\n124 g CO2/km (comb)\n'],
 ['\n', '\n', '\n99 g CO2/km (comb)\n'],
 ['\n', '\n', '\n128 g CO2/km (comb)\n'],
 ['\n', '\n', '\n119 g CO2/km (comb)\n'],
 ['\n', '\n', '\n140 g CO2/km (comb)\n'],
 ['\n', '\n', '\n106 g CO2/km (comb)\n'],
 ['\n', '\n', '\n92 g CO2/km (comb)\n'],
 ['\n', '\n', '\n107 g CO2/km (comb)\n'],
 ['\n', '\n', '\n114 g CO2/km (comb)\n'],
 ['\n', '\n', '\n1 (No sticker)\n'],
 ['\n', '\n', '\n102 g CO2/km (comb)\n'],
 ['\n', '\n', '\n103 g CO2/km (comb)\n'],
 ['\n', '\n', '\nEuro 6c\n'],
 ['\n', '\n', '\n117 g CO2/km (comb)\n'],
 ['\n', '\n', '\n139 g CO2/km (comb)\n'],
 ['\n', '\n', '\n127 g CO2/km (comb)\n'],
 ['\n', '\n', '\n141 g CO2/km (comb)\n'],
 ['\n', '\n', 

In [82]:
oto.drop("Full Service", axis = 1, inplace = True)

***

### ``'Comfort & Convenience'``

In [83]:
oto["\nComfort & Convenience\n"]

0                                                                                                                                                                                               [Air conditioning, Armrest, Automatic climate control, Cruise control, Electrical side mirrors, Hill Holder, Leather steering wheel, Light sensor, Multi-function steering wheel, Navigation system, Park Distance Control, Parking assist system sensors rear, Power windows, Rain sensor, Seat heating, Start-stop system]
1                                                                                                                                                                                                                                                                                                                                   [Air conditioning, Automatic climate control, Hill Holder, Leather steering wheel, Lumbar support, Parking assist system sensors rear, Power windows, Start-stop system

In [84]:
oto['\nComfort & Convenience\n'].value_counts(dropna = False)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                                                                                                                                                                                                                                                                                                     920
[Air conditioning, Electrical side mirrors, Hill Holder, Power windows]                                                                                                                                                                                                                                                                                                 216
[Air conditioning, Electrical side mirrors, Power windows]                                                                                                                                                                                                                      

In [85]:
oto['\nComfort & Convenience\n'].value_counts(dropna = False).head(50)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    920
[Air conditioning, Electrical side mirrors, Hill Holder, Power windows]                                                                                                                                                                                                                                                                                                                                                                                                                                      

In [86]:
oto['\nComfort & Convenience\n'].value_counts(dropna = False).tail(50)

TypeError: unhashable type: 'list'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1653, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'list'


[Air conditioning, Automatic climate control, Cruise control, Electrically adjustable seats, Electrical side mirrors, Heated steering wheel, Hill Holder, Leather steering wheel, Lumbar support, Multi-function steering wheel, Navigation system, Power windows, Rain sensor, Seat heating, Seat ventilation]                                                                                                                                        1
[Air conditioning, Armrest, Automatic climate control, Cruise control, Electrical side mirrors, Leather steering wheel, Multi-function steering wheel, Park Distance Control, Parking assist system sensors front, Parking assist system sensors rear, Split rear seats, Start-stop system]                                                                                                                                                            1
[Air conditioning, Armrest, Automatic climate control, Cruise control, Electrical side mirrors, Electric tailgate, Hea

> ### May be this column contains the most important data for prediction of the price.
> ### Because there are some variables that have list in list, so we have to see all 'combined' values the variables have. 
> ### We have to examine the unique values. We produce two function to do that.  First we convert ``NaN`` values to empty list than combine all values using ``set().update()`` syntax.

In [89]:
# We can't operate with NaNs. We have to examine the unique values of this column.

#uni = set()
#uni.update(np.nan)
#uni.update(["a", "b", "a"])
#list(uni)

In [123]:
# The new version of this data is a string. We have to convert its elments backto the list type.
# In order to use .get_dummies() method, we have to do that.

#oto['Comfort&Convenience'][0]

- **Defining a Function**

In [94]:
def unique_values(data, column_name):
    unique_vals = set()
    for rows in data[column_name]:
        unique_vals.update(rows)
    return list(unique_vals)

In [91]:
nans = oto['\nComfort & Convenience\n'].isnull()

In [92]:
oto.loc[nans, '\nComfort & Convenience\n']

9        NaN
30       NaN
53       NaN
168      NaN
191      NaN
        ... 
15791    NaN
15804    NaN
15805    NaN
15854    NaN
15864    NaN
Name: \nComfort & Convenience\n, Length: 920, dtype: object

In [96]:
def Nan_to_list(data, column_name):
    NaN_rows = data[column_name].isnull()
    data.loc[NaN_rows, column_name] = data.loc[NaN_rows, column_name].apply(lambda x : [])

In [97]:
Nan_to_list(oto, '\nComfort & Convenience\n')
unique_values(oto, '\nComfort & Convenience\n')

['Keyless central door lock',
 'Windshield',
 'Parking assist system camera',
 'Air conditioning',
 'Electrically heated windshield',
 'Rain sensor',
 'Power windows',
 'Electric tailgate',
 'Armrest',
 'Electrically adjustable seats',
 'Multi-function steering wheel',
 'Panorama roof',
 'Electric Starter',
 'Navigation system',
 'Split rear seats',
 'Hill Holder',
 'Heated steering wheel',
 'Sunroof',
 'Air suspension',
 'Parking assist system self-steering',
 'Seat heating',
 'Parking assist system sensors front',
 'Leather steering wheel',
 'Tinted windows',
 'Light sensor',
 'Leather seats',
 'Automatic climate control',
 'Auxiliary heating',
 'Cruise control',
 'Start-stop system',
 'Lumbar support',
 'Heads-up display',
 'Massage seats',
 'Seat ventilation',
 'Parking assist system sensors rear',
 'Electrical side mirrors',
 'Park Distance Control',
 'Wind deflector']

In [98]:
nxtins_list = [",".join(item).strip() if type(item) == list else item for item in oto['\nComfort & Convenience\n']]

In [99]:
oto["Comfort&Convenience"] = pd.DataFrame(nxtins_list)

In [100]:
oto['Comfort&Convenience'].value_counts(dropna=False)

                                                                                                                                                                                                                                                                                                                                                                                                                                                                     920
Air conditioning,Electrical side mirrors,Hill Holder,Power windows                                                                                                                                                                                                                                                                                                                                                                                                   216
Air conditioning,Electrical side mirrors,Power windows                                

In [103]:
oto[oto['Comfort&Convenience'] == ""]

url      make_model                                  short_description body_type  price             vat          km registration      hp                  Type Previous Owners Next Inspection Inspection new       Warranty null         Make             Model    Offer Number First Registration       Body Color      Paint Type                   Body Color Original Upholstery Nr. of Doors Nr. of Seats Model Code         Gearing Type    Displacement Cylinders  Drive chain                Fuel                                                               Consumption               CO2 Emission Emission Class \nComfort & Convenience\n \nEntertainment & Media\n      \nExtras\n \nSafety & Security\n                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          description Emission Label    Gears Country version Previous_Owner1 Previous_Owner  Inspection_New_Bool next_inspection paint_type country emission_label Comfort&Convenience
9                  https://www.autoscout24.com//offers/audi-a1-sportback-tfsi-ultra-95-s-tronic-amb-gasoline-white-e2c49a42-6aaa-4039-9af1-7546a20c110e         Audi A1              SPORTBACK TFSI ULTRA 95 S-TRONIC AMB.    Sedans  17990            None   16,103 km      06/2017   70 kW  [, Used, , Gasoline]             NaN             NaN            NaN   \n3 months\n   []     \nAudi\n      [\n, A1, \n]             NaN     [\n, 2017, \n]  [\n, White, \n]             NaN                           [\nBlanc\n]        NaN      [\n5\n]      [\n4\n]        NaN  [\n, Automatic, \n]    [\n999 cc\n]       NaN          NaN  [\n, Gasoline, \n]                                                                       NaN                        NaN            NaN                        []                       NaN             NaN                   NaN  [\nClim automatique,Ecran multifonction couleur,Capteur de pluie,Capteur de luminosité,Contrôle élect. de la pression des pneus,Régulateur de vitesse,EBD,ESP,8 Haut parleurs,ABS,Airbag conducteur,Airbag passager,Airbags latéraux avant,Airbags rideaux AV et AR,Antidémarrage électronique,Antipatinage,Appui-tête conducteur réglable hauteur,Appui-tête passager réglable en hauteur,Arrêt et redémarrage auto. du moteur,Bacs de portes avant,Banquette 1/3-2/3,Banquette AR rabattable,Boite à gant ferm...            NaN  [\n7\n]             NaN             NaN              1                    0             NaT        NaN     NaN            NaN                    
30                            https://www.autoscout24.com//offers/audi-a1-sportback-1-4tdi-attraction-diesel-black-be5e7705-5e6e-4ca0-a7df-fc6153d6e384         Audi A1                        Sportback 1.4TDI Attraction   Compact  11290            None   38,400 km      06/2016   66 kW    [, Used, , Diesel]             NaN             NaN            NaN  \n12 months\n   []     \nAudi\n      [\n, A1, \n]   [\n2780690\n]     [\n, 2016, \n]  [\n, Black, \n]             NaN                           [\nNegro\n]        NaN      [\n5\n]      [\n5\n]        NaN     [\n, Manual, \n]  [\n1,422 cc\n]         3  [\nfront\n]    [\n, Diesel, \n]        [[3 l/100 km (comb)], [3 l/100 km (city)], [3 l/100 km (country)]]                        NaN            NaN                        []                       NaN             NaN                   NaN  [\nVEHÍCULO NACIONAL, PROCEDE DE CONCESIONARIO OFICIAL , KMTS CERTIFICADOS CON LIBRO DE REVISIONES EN CONCESIONARIO OFICIAL, MUY BUEN ESTADO GENERAL Y MECANICO (21% IVA INCLUIDO EN EL PRECIO), AUX, SROP&GO, 5 PLAZAS, CIERRE CENTRALIZADO , ESPEJOS AUT..., ACEPTAMOS VEHICULO COMO PAGO , 

In [104]:
oto['Comfort&Convenience'].isnull().sum()

0

In [105]:
[np.nan if oto['Comfort&Convenience'].iloc[i] == "" else oto['Comfort&Convenience'].iloc[i] for i in range(len(oto['Comfort&Convenience']))] 

['Air conditioning,Armrest,Automatic climate control,Cruise control,Electrical side mirrors,Hill Holder,Leather steering wheel,Light sensor,Multi-function steering wheel,Navigation system,Park Distance Control,Parking assist system sensors rear,Power windows,Rain sensor,Seat heating,Start-stop system',
 'Air conditioning,Automatic climate control,Hill Holder,Leather steering wheel,Lumbar support,Parking assist system sensors rear,Power windows,Start-stop system,Tinted windows',
 'Air conditioning,Cruise control,Electrical side mirrors,Hill Holder,Leather steering wheel,Multi-function steering wheel,Navigation system,Park Distance Control,Parking assist system sensors front,Parking assist system sensors rear,Power windows,Seat heating,Start-stop system',
 'Air suspension,Armrest,Auxiliary heating,Electrical side mirrors,Heads-up display,Hill Holder,Leather steering wheel,Light sensor,Lumbar support,Multi-function steering wheel,Navigation system,Power windows,Rain sensor,Seat heating,Sp

In [106]:
oto["Comfort&Convenience"] = [np.nan if oto['Comfort&Convenience'].iloc[i] == "" else oto['Comfort&Convenience'].iloc[i] for i in range(len(oto['Comfort&Convenience']))]

In [107]:
oto["Comfort&Convenience"].value_counts(dropna = False)

NaN                                                                                                                                                                                                                                                                                                                                                                                                                                                                  920
Air conditioning,Electrical side mirrors,Hill Holder,Power windows                                                                                                                                                                                                                                                                                                                                                                                                   216
Air conditioning,Electrical side mirrors,Power windows                                

In [108]:
fill_most(oto, "make_model", "Comfort&Convenience")

In [109]:
oto['Comfort&Convenience'].isnull().sum()

0

In [110]:
oto['Comfort&Convenience'].value_counts()

Air conditioning,Electrical side mirrors,Hill Holder,Power windows                                                                                                                                                                                                                                                                                                                                                                                                                             389
Air conditioning,Armrest,Automatic climate control,Cruise control,Electrically adjustable seats,Electrical side mirrors,Electric tailgate,Heated steering wheel,Hill Holder,Keyless central door lock,Leather steering wheel,Light sensor,Lumbar support,Multi-function steering wheel,Navigation system,Park Distance Control,Parking assist system camera,Parking assist system sensors front,Parking assist system sensors rear,Power windows,Rain sensor,Seat heating,Start-stop system    259
Air conditioning,Cruise control,El

In [114]:
"Windshield" in oto["Comfort&Convenience"][0]

False

In [112]:
oto["Comfort&Convenience"].str.get_dummies(sep = ",")

Air conditioning  Air suspension  Armrest  Automatic climate control  Auxiliary heating  Cruise control  Electric Starter  Electric tailgate  Electrical side mirrors  Electrically adjustable seats  Electrically heated windshield  Heads-up display  Heated steering wheel  Hill Holder  Keyless central door lock  Leather seats  Leather steering wheel  Light sensor  Lumbar support  Massage seats  Multi-function steering wheel  Navigation system  Panorama roof  Park Distance Control  Parking assist system camera  Parking assist system self-steering  Parking assist system sensors front  Parking assist system sensors rear  Power windows  Rain sensor  Seat heating  Seat ventilation  Split rear seats  Start-stop system  Sunroof  Tinted windows  Wind deflector  Windshield
0                     1               0        1                          1                  0               1                 0                  0                        1                              0                               0                 0                      0            1                          0              0                       1             1               0              0                              1                  1              0                      1                             0                                    0                                    0                                   1              1            1             1                 0                 0                  1        0               0               0           0
1                     1               0        0                          1                  0               0                 0                  0                        0                              0                               0                 0                      0            1                          0              0                       1             0               1              0                              0                  0              0                      0                             0                                    0                                    0                                   1              1            0             0                 0                 0                  1        0               1               0           0
2                     1               0        0                          0                  0               1                 0                  0                        1                              0                               0                 0                      0            1                          0              0                       1             0               0              0                              1                  1              0                      1                             0                                    0                                    1                                   1              1            0             1                 0                 0                  1        0               0               0           0
3                     0               1        1                          0                  1               0                 0                  0                        1                              0                               0                 1                      0            1                          0              0                       1             1               1              0                              1                  1              0                      0                             0                                    0                                    0                                   0              1            1             1                 0                 1                  1        0               0               0           0
4                     1               0        1                          1                  0               0  

***

### ``'\nEntertainment & Media\n'``

***

### ``'\nExtras\n'``

***

### '\nSafety & Security\n'

***

### ``'Consumption'``

#### There are three tpyes of value which can effect the result. 
 - Type 1-one list and three sub list (comb,city,country), like= ``[[3.8 l/100 km (comb)], [4.3 l/100 km (city)], [3.5 l/100 km (country)]]``  
 - Type 2-one list and no sub list (comb,city,country), like = ``[\n, 7.4 l/100 km (comb), \n, 9.2 l/100 km (city), \n, 6.3 l/100 km (country), \n]``  
 - Type 3-one list and three sub lists some of which are empty, like = ``[[], [4.6 l/100 km (city)], [5.7 l/100 km (country)]]`` or ``[[4.2 l/100 km (comb)], [], []]`` or `[[], [], [5.1 l/100 km (country)]]`
 - So, we use ``str.extract()`` to select numeric consumption(comb) in different ways below.

- There are 850 values in which only 'one' list (the value of ``"\n"`` indicates that). Consists of no sub list. 

- There are 127 values is ``'[]'`` which means there is no ``comb`` consumption. Only have ``city`` or ``country`` or both.

- There are still 1906 ``NaN``s.

- If consumption is two digits, Extraction function takes only first one. 
   - eg.: ``40`` → ``4``
   - eg.: ``33`` → ``3``

- Now, lets go into the other types. Let's try ``str[1]``

- Let's take a look at the "``country``"

***

### ``'Gearing Type'``

***

### ``'Drive chain'``

- We can handle with missig values or we can drop it. Let's take a look at this issue later.

***

### ``'Model Code'``

***

## Now we start to analyze the variables which has below %35 percentage of missing values.

### ``'Body Color'``

***

### ``'Body Color Original'``

***

### ``'CO2 Emission'``

- let's fill missing values using median.

***

### ``'Displacement'``

***

### ``'Emission Class'``

***

### ``'First Registration'``

- We can transform this feature into an "age" column. Let's create a new column called ``"register_age"``

***

### ``'Fuel'``

***

### ``'Gears'``

***

### ``'Make'``

***

### ``'Model'``

***

### ``'Nr. of Doors'``

***

### 'Nr. of Seats'

***

### 'Offer Number'

***

### 'Type'

***

### 'Upholstery'

***

### 'Warranty'

- There are several numeric values in the same line
- In order to select months from among the other numeric values, we should use the syntax below.
- Firstly, we will select ``"dd months"``
- Secondly, we replace all chars other than numeric

- It has too many missing values so, it will not effect the result if we drop it.

***

### ``'description'``

***

### 'hp'

***

### ``'km'``

- There are "-" values which means ``NaN``s. We can not accept these values as 0 km.

***

### ``'make_model' ``

***

### ``'null'``

***

### ``'registration'``

***

### ``'short_description'``

***

### ``'url'``

***

### ``'vat'``

***

### ``'price'``

- This is our target variable which we try to predict..

### Now, it is time to drop repeated columns we have.

## End of this Project. Next step will be :
- Handling with Missing Values and Outliers,
- Continue to clean DataSet.
- Make the DataSet ready to go into ML Models.